oidc/clientassertion/client_assertion.go

```diff
@@ -0,0 +1,201 @@
+}
+
+// Validate validates the expected fields
+func (j *JWT) Validate() error {
+	var errs []error
+	if j.genID == nil {
+		errs = append(errs, ErrMissingFuncIDGenerator)
+	}
+	if j.now == nil {
+		errs = append(errs, ErrMissingFuncNow)
+	}
+	// bail early if any internal func errors
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	if j.clientID == "" {
+		errs = append(errs, ErrMissingClientID)
+	}
+	if len(j.audience) == 0 {
+		errs = append(errs, ErrMissingAudience)
+	}
+	if j.alg == "" {
+		errs = append(errs, ErrMissingAlgorithm)
+	}
+	if j.key == nil && j.secret == "" {
+		errs = append(errs, ErrMissingKeyOrSecret)
+	}
+	if j.key != nil && j.secret != "" {
+		errs = append(errs, ErrBothKeyAndSecret)
+	}
+	// if any of those fail, we have no hope.
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	// finally, make sure Serialize() works; we can't pre-validate everything,
+	// and this whole thing is useless if it can't Serialize()
+	if _, err := j.Serialize(); err != nil {
+		return fmt.Errorf("serialization error during validate: %w", err)
+	}
+
+	return nil
+}
+
+// Serialize returns a signed JWT string
+func (j *JWT) Serialize() (string, error) {
+	builder, err := j.builder()
+	if err != nil {
+		return "", err
+	}
+	token, err := builder.Serialize()
+	if err != nil {
+		return "", fmt.Errorf("failed to sign token: %w", err)
+	}
+	return token, nil
+}
+
+func (j *JWT) builder() (jwt.Builder, error) {
+	signer, err := j.signer()
+	if err != nil {
+		return nil, err
+	}
+	id, err := j.genID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate token id: %w", err)
+	}
+	claims := j.claims(id)
+	return jwt.Signed(signer).Claims(claims), nil
+}
+
+func (j *JWT) signer() (jose.Signer, error) {
+	sKey := jose.SigningKey{
+		Algorithm: j.alg,
+	}
+
+	// Validate() ensures these are mutually exclusive
+	if j.secret != "" {
+		sKey.Key = []byte(j.secret)
+	}
+	if j.key != nil {
+		sKey.Key = j.key
+	}
+
+	sOpts := &jose.SignerOptions{
+		ExtraHeaders: make(map[jose.HeaderKey]interface{}, len(j.headers)),
+	}
+	// note: extra headers can override "kid"
+	for k, v := range j.headers {
+		sOpts.ExtraHeaders[jose.HeaderKey(k)] = v
+	}
+
+	signer, err := jose.NewSigner(sKey, sOpts.WithType("JWT"))
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrCreatingSigner, err)
+	}
+	return signer, nil
+}
+
+func (j *JWT) claims(id string) *jwt.Claims {
+	now := j.now().UTC()
+	return &jwt.Claims{
+		Issuer:    j.clientID,
+		Subject:   j.clientID,
+		Audience:  j.audience,
+		Expiry:    jwt.NewNumericDate(now.Add(5 * time.Minute)),
+		NotBefore: jwt.NewNumericDate(now.Add(-1 * time.Second)),
+		IssuedAt:  jwt.NewNumericDate(now),
+		ID:        id,
+	}
+}
+
+// Serializer is the primary interface impelmented by JWT.
+type Serializer interface {
+	Serialize() (string, error)
+}
+
+// ensure JWT implements Serializer, which is accepted by the oidc option
+// oidc.WithClientAssertionJWT.
+var _ Serializer = &JWT{}
```

Comment:
I don't think we really need this interface, or at the very least we don't need to export it. In practice we should notice very quickly if `JWT` is no longer usable with `oid.JWTSerializer`.